src/class-bulk-task.php

```diff
@@ -0,0 +1,238 @@
+<?php
+/**
+ * Alley\WP_Bulk_Task: Bulk_Task class
+ *
+ * @package alleyinteractive/wp-bulk-task
+ */
+
+namespace Alley\WP_Bulk_Task;
+
+use Alley\WP_Bulk_Task\Progress\Progress;
+use WP_Query;
+
+/**
+ * A class that provides performant bulk task functionality.
+ *
+ * @package alleyinteractive/wp-bulk-task
+ */
+class Bulk_Task {
+
+	/**
+	 * The cursor associated with this bulk task.
+	 *
+	 * @var Cursor
+	 */
+	public Cursor $cursor;
+
+	/**
+	 * Store the max post ID from the posts table so we know when we are done.
+	 *
+	 * @var int
+	 */
+	protected int $max_id;
+
+	/**
+	 * Store the last processed post ID for bulk task pagination.
+	 *
+	 * @var int
+	 */
+	protected int $min_id;
+
+	/**
+	 * Store the current WP_Query object hash for bulk tasks.
+	 *
+	 * @var string
+	 */
+	protected string $object_hash;
+
+	/**
+	 * The slice of IDs in the database within which to look for matching posts.
+	 * By keeping this number higher than the posts_per_page, but smaller than
+	 * the total number of results in very large databases, it enables the right
+	 * balance of performant lookups with moving through the database quickly when
+	 * there are no matches in a particular slice of results.
+	 */
+	const STEPPING = 10000;
```

Comment:
I'd advise that this be a public property vs a const, so someone could override it. and it should default to either 100x posts_per_page or `max($this->stepping, $args['posts_per_page'])`. as of right now, if one sets `posts_per_page` to 50000, each query will still be limited to 10000.

Reply:
In my testing, I found that 10x posts_per_page was too unreliable (the performance of looking for 100 posts in 1,000 was not nearly as good as looking in 10,000, and I didn't want to do 100x in case posts_per_page was set really high, like 1,000 or 2,000). I like the suggestion of making it a public property so it's overridable and choosing the max between stepping and posts per page—I'll apply that.